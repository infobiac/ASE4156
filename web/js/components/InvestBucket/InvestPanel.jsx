// @flow
import React from 'react';
import Dialog, {
  DialogContent,
  DialogTitle,
  DialogActions,
} from 'material-ui/Dialog';
import Button from 'material-ui/Button';
import Table, { TableBody, TableCell, TableHead, TableRow } from 'material-ui/Table';
import Slider from 'rc-slider';

import StockGraph from '../StockGraph/StockGraph';

type Props = {
  available: number,
  bucket: {
    name: string,
    value: number,
    ownedAmount: number,
    history: Array<{date: Date, value: number}>
  },
  investFunc: number => void,
  cancelFunc: () => void,
}
type State = {
  investedAmount: number,
}

class InvestPanel extends React.Component<Props, State> {
  constructor() {
    super();
    this.state = {
      investedAmount: 0,
    };
  }
  render() {
    const values = [{
      name: this.props.bucket.name,
      data: this.props.bucket.history,
    }];
    const ownedQuantity = this.props.bucket.ownedAmount;
    const additionalQuantity = this.state.investedAmount / this.props.bucket.value;
    const ownedValue = this.props.bucket.ownedAmount * this.props.bucket.value;
    const additionalValue = this.state.investedAmount;
    const totalValue = ownedValue + additionalValue;

    return (
      <Dialog open>
        <DialogTitle>Invest into bucket</DialogTitle>
        <DialogContent>
          <StockGraph
            id="chart"
            title="Test Chart"
            quotes={values}
          />
          <Slider
            value={totalValue}
            id="invest-amount-slider"
            onChange={investedAmount => this.setState(() => ({
              investedAmount: investedAmount - ownedValue,
            }))}
            max={this.props.available + ownedValue}
            step={0.01}
          />
          <Table>
            <TableHead>
              <TableRow>
                <TableCell padding="dense">Name</TableCell>
                <TableCell padding="dense">Quantity</TableCell>
                <TableCell padding="dense">Value</TableCell>
                <TableCell padding="dense">Total Value</TableCell>
              </TableRow>
            </TableHead>
            <TableBody>
              <TableRow>
                <TableCell padding="dense">{this.props.bucket.name} (Owned)</TableCell>
                <TableCell padding="dense">{ownedQuantity.toFixed(2)}</TableCell>
                <TableCell padding="dense">{this.props.bucket.value.toFixed(2)}</TableCell>
                <TableCell padding="dense">{ownedValue.toFixed(2)}</TableCell>
              </TableRow>
              <TableRow>
                <TableCell padding="dense">
                  {this.props.bucket.name}
                  {
                    additionalValue !== 0
                      ? <span>({additionalValue > 0 ? 'Buy' : 'Sell'})</span>
                      : null
                  }
                </TableCell>
                <TableCell padding="dense">{additionalQuantity.toFixed(2)}</TableCell>
                <TableCell padding="dense">{this.props.bucket.value.toFixed(2)}</TableCell>
                <TableCell padding="dense">{additionalValue.toFixed(2)}</TableCell>
              </TableRow>
              <TableRow>
                <TableCell padding="dense">Available</TableCell>
                <TableCell padding="dense" />
                <TableCell padding="dense" />
                <TableCell padding="dense">{(this.props.available - this.state.investedAmount).toFixed(2)}</TableCell>
              </TableRow>
            </TableBody>
          </Table>
        </DialogContent>
        <DialogActions>
          <Button id="cancel" onClick={this.props.cancelFunc}>Cancel</Button>
          <Button id="save" onClick={() => this.props.investFunc(additionalQuantity)}>
            Invest
          </Button>
        </DialogActions>
      </Dialog>);
  }
}

export default InvestPanel;
