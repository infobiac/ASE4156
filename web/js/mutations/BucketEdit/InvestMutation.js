// @flow
import { graphql, commitMutation } from 'react-relay';

import type {
  RecordSourceSelectorProxy,
  RelayEnvironment,
  SelectorData,
} from 'react-relay';

import type {
  InvestMutationMutationVariables,
  InvestMutationMutationResponse,
} from './__generated__/InvestMutationMutation.graphql';

const mutation = graphql`
  mutation InvestMutationMutation(
    $quantity: Float!,
    $tradingAccId: ID!,
    $bucketId: ID!,
  ) {
    invest(
      quantity: $quantity,
      tradingAccId: $tradingAccId,
      bucketId: $bucketId,
    ) {
      tradingAccount {
        id
        availableCash
      }
      bucket {
        id
        ownedAmount
      }
    }
  }
`;

export default (
  updater?: ?(store: RecordSourceSelectorProxy, data: SelectorData) => void,
  optimisticUpdater?: ?(store: RecordSourceSelectorProxy) => void,
  onCompleted?: ?(response: ?InvestMutationMutationResponse, errors: ?[Error]) => void,
) => (environment: RelayEnvironment) => (
  variables: InvestMutationMutationVariables,
  optimisticResponse?: ?InvestMutationMutationResponse,
) => {
  commitMutation(environment, {
    mutation,
    variables,
    optimisticResponse,
    onCompleted,
  });
};
