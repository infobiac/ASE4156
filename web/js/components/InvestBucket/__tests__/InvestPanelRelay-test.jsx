import React from 'react';
import { mount } from 'enzyme';
import {
  Environment,
  RecordSource,
  Store,
} from 'relay-runtime';
import PropTypes from 'prop-types';
import { routerShape } from 'found';

import type { Node } from 'react';

import InvestPanel from '../InvestPanel';
import InvestPanelRelay from '../InvestPanelRelay';
import InvestMutation from '../../../mutations/BucketEdit/InvestMutation';

jest.mock(
  '../InvestPanel',
  () => jest.fn(() => <div />),
);
jest.mock(
  '../../../mutations/BucketEdit/InvestMutation',
  () => {
    const inner1 = jest.fn();
    const inner2 = jest.fn(() => inner1);
    const outer = jest.fn(() => inner2);
    return outer;
  },
);

const ctx = () => {
  const source = new RecordSource();
  const store = new Store(source);
  const network = jest.fn();
  const relay = {
    environment: new Environment({
      network,
      store,
    }),
    variables: {},
  };
  return ({
    errorDisplay: jest.fn(),
    router: {
      push: jest.fn(),
      replace: jest.fn(),
      go: jest.fn(),
      createHref: jest.fn(),
      createLocation: jest.fn(),
      isActive: jest.fn(),
      matcher: {
        match: jest.fn(),
        getRoutes: jest.fn(),
        isActive: jest.fn(),
        format: jest.fn(),
      },
      addTransitionHook: jest.fn(),
    },
    relay,
  });
};

class CtxProvider extends React.Component<{children: () => Node, ctx: any}> {
  static childContextTypes = {
    errorDisplay: PropTypes.func.isRequired,
    router: routerShape.isRequired,
    relay: PropTypes.any.isRequired,
  }
  getChildContext() {
    return this.props.ctx;
  }
  render() {
    return (
      <div>
        {this.props.children(this.props.ctx.relay)}
      </div>
    );
  }
}

describe('InvestPanelRelay', () => {
  it('exists', () => {
    expect(InvestPanelRelay).toBeTruthy();
  });

  it('renders', () => {
    const context = ctx();
    const comp = mount((
      <CtxProvider ctx={context}>
        {
          con => (
            <InvestPanelRelay
              id="panel"
              relay={con.relay}
              bucket={{
                __id: '1',
                __fragments: { InvestPanelRelay_bucket: {} },
                id: '1',
                name: 'Some name',
                value: 33.3,
                ownedAmount: 7,
                history: [{ date: '2017-07-07', value: 55.3 }],
              }}
              profile={{
                __id: '2',
                __fragments: { InvestPanelRelay_profile: {} },
                id: '2',
                selectedAcc: {
                  availableCash: 1234.56,
                  id: '3',
                },
              }}
            />
          )
        }
      </CtxProvider>
    ));
    expect(comp).toMatchSnapshot();
    // Cancel closes
    InvestPanel.mock.calls[0][0].cancelFunc();
    expect(context.router.replace.mock.calls.length).toEqual(1);
    expect(context.router.replace.mock.calls[0][0]).toEqual('/home');
    // Save
    InvestPanel.mock.calls[0][0].investFunc(3);
    expect(InvestMutation.mock.calls.length).toEqual(1);
    expect(InvestMutation.mock.calls[0]).toMatchSnapshot();
    expect(InvestMutation().mock.calls.length).toEqual(1);
    expect(InvestMutation().mock.calls[0]).toMatchSnapshot();
    expect(InvestMutation()().mock.calls.length).toEqual(1);
    expect(InvestMutation()().mock.calls[0]).toMatchSnapshot();
    InvestPanel.mock.calls[0][0].investFunc(-3);
    InvestMutation.mock.calls[0][2]({}, null);
    InvestMutation.mock.calls[InvestMutation.mock.calls.length - 1][2]({}, null);
    InvestMutation.mock.calls[0][2](null, [Error('Test')]);
    expect(context.errorDisplay.mock.calls).toMatchSnapshot();
  });
});
