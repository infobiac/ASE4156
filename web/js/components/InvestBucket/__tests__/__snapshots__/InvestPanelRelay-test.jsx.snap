// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvestPanelRelay renders 1`] = `
<CtxProvider
  ctx={
    Object {
      "errorDisplay": [Function],
      "relay": Object {
        "environment": RelayModernEnvironment {
          "__setNet": [Function],
          "_debugger": EnvironmentDebugger {
            "_envIsDirty": false,
            "_environment": [Circular],
            "_id": "RelayModernEnvironment1",
            "_isRecordingMutationEvents": false,
            "_recordedMutationEvents": Array [],
          },
          "_network": [Function],
          "_publishQueue": RelayPublishQueue {
            "_appliedOptimisticUpdates": Set {},
            "_backup": Object {},
            "_handlerProvider": [Function],
            "_pendingBackupRebase": false,
            "_pendingData": Set {},
            "_pendingOptimisticUpdates": Set {},
            "_pendingUpdaters": Set {},
            "_store": RelayMarkSweepStore {
              "_hasScheduledGC": false,
              "_index": 0,
              "_recordSource": Object {},
              "_roots": Map {},
              "_subscriptions": Set {
                Object {
                  "callback": [Function],
                  "snapshot": Object {
                    "data": undefined,
                    "dataID": "1",
                    "node": Object {
                      "argumentDefinitions": Array [],
                      "kind": "Fragment",
                      "metadata": null,
                      "name": "InvestPanelRelay_bucket",
                      "selections": Array [
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "id",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "name",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "value",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": null,
                          "kind": "ScalarField",
                          "name": "ownedAmount",
                          "storageKey": null,
                        },
                        Object {
                          "alias": null,
                          "args": Array [
                            Object {
                              "kind": "Literal",
                              "name": "count",
                              "type": "Int",
                              "value": 30,
                            },
                          ],
                          "concreteType": "GDataPoint",
                          "kind": "LinkedField",
                          "name": "history",
                          "plural": true,
                          "selections": Array [
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "date",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "value",
                              "storageKey": null,
                            },
                          ],
                          "storageKey": "history{\\"count\\":30}",
                        },
                      ],
                      "type": "GInvestmentBucket",
                    },
                    "seenRecords": Object {
                      "1": undefined,
                    },
                    "variables": Object {},
                  },
                },
                Object {
                  "callback": [Function],
                  "snapshot": Object {
                    "data": undefined,
                    "dataID": "2",
                    "node": Object {
                      "argumentDefinitions": Array [],
                      "kind": "Fragment",
                      "metadata": null,
                      "name": "InvestPanelRelay_profile",
                      "selections": Array [
                        Object {
                          "alias": null,
                          "args": null,
                          "concreteType": "GTradingAccount",
                          "kind": "LinkedField",
                          "name": "selectedAcc",
                          "plural": false,
                          "selections": Array [
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "availableCash",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "id",
                              "storageKey": null,
                            },
                          ],
                          "storageKey": null,
                        },
                      ],
                      "type": "GProfile",
                    },
                    "seenRecords": Object {
                      "2": undefined,
                    },
                    "variables": Object {},
                  },
                },
              },
              "_updatedRecordIDs": Object {},
            },
          },
          "_store": RelayMarkSweepStore {
            "_hasScheduledGC": false,
            "_index": 0,
            "_recordSource": Object {},
            "_roots": Map {},
            "_subscriptions": Set {
              Object {
                "callback": [Function],
                "snapshot": Object {
                  "data": undefined,
                  "dataID": "1",
                  "node": Object {
                    "argumentDefinitions": Array [],
                    "kind": "Fragment",
                    "metadata": null,
                    "name": "InvestPanelRelay_bucket",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "id",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "name",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "value",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "ownedAmount",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": Array [
                          Object {
                            "kind": "Literal",
                            "name": "count",
                            "type": "Int",
                            "value": 30,
                          },
                        ],
                        "concreteType": "GDataPoint",
                        "kind": "LinkedField",
                        "name": "history",
                        "plural": true,
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "date",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "value",
                            "storageKey": null,
                          },
                        ],
                        "storageKey": "history{\\"count\\":30}",
                      },
                    ],
                    "type": "GInvestmentBucket",
                  },
                  "seenRecords": Object {
                    "1": undefined,
                  },
                  "variables": Object {},
                },
              },
              Object {
                "callback": [Function],
                "snapshot": Object {
                  "data": undefined,
                  "dataID": "2",
                  "node": Object {
                    "argumentDefinitions": Array [],
                    "kind": "Fragment",
                    "metadata": null,
                    "name": "InvestPanelRelay_profile",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": null,
                        "concreteType": "GTradingAccount",
                        "kind": "LinkedField",
                        "name": "selectedAcc",
                        "plural": false,
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "availableCash",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "id",
                            "storageKey": null,
                          },
                        ],
                        "storageKey": null,
                      },
                    ],
                    "type": "GProfile",
                  },
                  "seenRecords": Object {
                    "2": undefined,
                  },
                  "variables": Object {},
                },
              },
            },
            "_updatedRecordIDs": Object {},
          },
          "configName": undefined,
          "unstable_internal": Object {
            "areEqualSelectors": [Function],
            "createFragmentSpecResolver": [Function],
            "createOperationSelector": [Function],
            "getDataIDsFromObject": [Function],
            "getFragment": [Function],
            "getOperation": [Function],
            "getSelector": [Function],
            "getSelectorList": [Function],
            "getSelectorsFromObject": [Function],
            "getVariablesFromObject": [Function],
          },
        },
        "variables": Object {},
      },
      "router": Object {
        "addTransitionHook": [Function],
        "createHref": [Function],
        "createLocation": [Function],
        "go": [Function],
        "isActive": [Function],
        "matcher": Object {
          "format": [Function],
          "getRoutes": [Function],
          "isActive": [Function],
          "match": [Function],
        },
        "push": [Function],
        "replace": [Function],
      },
    }
  }
>
  <div>
    <Relay(InvestPanelRelay)
      bucket={
        Object {
          "__fragments": Object {
            "InvestPanelRelay_bucket": Object {},
          },
          "__id": "1",
          "history": Array [
            Object {
              "date": "2017-07-07",
              "value": 55.3,
            },
          ],
          "id": "1",
          "name": "Some name",
          "ownedAmount": 7,
          "value": 33.3,
        }
      }
      id="panel"
      profile={
        Object {
          "__fragments": Object {
            "InvestPanelRelay_profile": Object {},
          },
          "__id": "2",
          "id": "2",
          "selectedAcc": Object {
            "availableCash": 1234.56,
            "id": "3",
          },
        }
      }
    >
      <InvestPanelRelay
        bucket={
          Object {
            "__fragments": Object {
              "InvestPanelRelay_bucket": Object {},
            },
            "__id": "1",
            "history": Array [
              Object {
                "date": "2017-07-07",
                "value": 55.3,
              },
            ],
            "id": "1",
            "name": "Some name",
            "ownedAmount": 7,
            "value": 33.3,
          }
        }
        id="panel"
        profile={
          Object {
            "__fragments": Object {
              "InvestPanelRelay_profile": Object {},
            },
            "__id": "2",
            "id": "2",
            "selectedAcc": Object {
              "availableCash": 1234.56,
              "id": "3",
            },
          }
        }
        relay={
          Object {
            "environment": RelayModernEnvironment {
              "__setNet": [Function],
              "_debugger": EnvironmentDebugger {
                "_envIsDirty": false,
                "_environment": [Circular],
                "_id": "RelayModernEnvironment1",
                "_isRecordingMutationEvents": false,
                "_recordedMutationEvents": Array [],
              },
              "_network": [Function],
              "_publishQueue": RelayPublishQueue {
                "_appliedOptimisticUpdates": Set {},
                "_backup": Object {},
                "_handlerProvider": [Function],
                "_pendingBackupRebase": false,
                "_pendingData": Set {},
                "_pendingOptimisticUpdates": Set {},
                "_pendingUpdaters": Set {},
                "_store": RelayMarkSweepStore {
                  "_hasScheduledGC": false,
                  "_index": 0,
                  "_recordSource": Object {},
                  "_roots": Map {},
                  "_subscriptions": Set {
                    Object {
                      "callback": [Function],
                      "snapshot": Object {
                        "data": undefined,
                        "dataID": "1",
                        "node": Object {
                          "argumentDefinitions": Array [],
                          "kind": "Fragment",
                          "metadata": null,
                          "name": "InvestPanelRelay_bucket",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "id",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "name",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "value",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": null,
                              "kind": "ScalarField",
                              "name": "ownedAmount",
                              "storageKey": null,
                            },
                            Object {
                              "alias": null,
                              "args": Array [
                                Object {
                                  "kind": "Literal",
                                  "name": "count",
                                  "type": "Int",
                                  "value": 30,
                                },
                              ],
                              "concreteType": "GDataPoint",
                              "kind": "LinkedField",
                              "name": "history",
                              "plural": true,
                              "selections": Array [
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "date",
                                  "storageKey": null,
                                },
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "value",
                                  "storageKey": null,
                                },
                              ],
                              "storageKey": "history{\\"count\\":30}",
                            },
                          ],
                          "type": "GInvestmentBucket",
                        },
                        "seenRecords": Object {
                          "1": undefined,
                        },
                        "variables": Object {},
                      },
                    },
                    Object {
                      "callback": [Function],
                      "snapshot": Object {
                        "data": undefined,
                        "dataID": "2",
                        "node": Object {
                          "argumentDefinitions": Array [],
                          "kind": "Fragment",
                          "metadata": null,
                          "name": "InvestPanelRelay_profile",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "args": null,
                              "concreteType": "GTradingAccount",
                              "kind": "LinkedField",
                              "name": "selectedAcc",
                              "plural": false,
                              "selections": Array [
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "availableCash",
                                  "storageKey": null,
                                },
                                Object {
                                  "alias": null,
                                  "args": null,
                                  "kind": "ScalarField",
                                  "name": "id",
                                  "storageKey": null,
                                },
                              ],
                              "storageKey": null,
                            },
                          ],
                          "type": "GProfile",
                        },
                        "seenRecords": Object {
                          "2": undefined,
                        },
                        "variables": Object {},
                      },
                    },
                  },
                  "_updatedRecordIDs": Object {},
                },
              },
              "_store": RelayMarkSweepStore {
                "_hasScheduledGC": false,
                "_index": 0,
                "_recordSource": Object {},
                "_roots": Map {},
                "_subscriptions": Set {
                  Object {
                    "callback": [Function],
                    "snapshot": Object {
                      "data": undefined,
                      "dataID": "1",
                      "node": Object {
                        "argumentDefinitions": Array [],
                        "kind": "Fragment",
                        "metadata": null,
                        "name": "InvestPanelRelay_bucket",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "id",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "name",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "value",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": null,
                            "kind": "ScalarField",
                            "name": "ownedAmount",
                            "storageKey": null,
                          },
                          Object {
                            "alias": null,
                            "args": Array [
                              Object {
                                "kind": "Literal",
                                "name": "count",
                                "type": "Int",
                                "value": 30,
                              },
                            ],
                            "concreteType": "GDataPoint",
                            "kind": "LinkedField",
                            "name": "history",
                            "plural": true,
                            "selections": Array [
                              Object {
                                "alias": null,
                                "args": null,
                                "kind": "ScalarField",
                                "name": "date",
                                "storageKey": null,
                              },
                              Object {
                                "alias": null,
                                "args": null,
                                "kind": "ScalarField",
                                "name": "value",
                                "storageKey": null,
                              },
                            ],
                            "storageKey": "history{\\"count\\":30}",
                          },
                        ],
                        "type": "GInvestmentBucket",
                      },
                      "seenRecords": Object {
                        "1": undefined,
                      },
                      "variables": Object {},
                    },
                  },
                  Object {
                    "callback": [Function],
                    "snapshot": Object {
                      "data": undefined,
                      "dataID": "2",
                      "node": Object {
                        "argumentDefinitions": Array [],
                        "kind": "Fragment",
                        "metadata": null,
                        "name": "InvestPanelRelay_profile",
                        "selections": Array [
                          Object {
                            "alias": null,
                            "args": null,
                            "concreteType": "GTradingAccount",
                            "kind": "LinkedField",
                            "name": "selectedAcc",
                            "plural": false,
                            "selections": Array [
                              Object {
                                "alias": null,
                                "args": null,
                                "kind": "ScalarField",
                                "name": "availableCash",
                                "storageKey": null,
                              },
                              Object {
                                "alias": null,
                                "args": null,
                                "kind": "ScalarField",
                                "name": "id",
                                "storageKey": null,
                              },
                            ],
                            "storageKey": null,
                          },
                        ],
                        "type": "GProfile",
                      },
                      "seenRecords": Object {
                        "2": undefined,
                      },
                      "variables": Object {},
                    },
                  },
                },
                "_updatedRecordIDs": Object {},
              },
              "configName": undefined,
              "unstable_internal": Object {
                "areEqualSelectors": [Function],
                "createFragmentSpecResolver": [Function],
                "createOperationSelector": [Function],
                "getDataIDsFromObject": [Function],
                "getFragment": [Function],
                "getOperation": [Function],
                "getSelector": [Function],
                "getSelectorList": [Function],
                "getSelectorsFromObject": [Function],
                "getVariablesFromObject": [Function],
              },
            },
          }
        }
      >
        <mockConstructor
          available={1234.56}
          bucket={
            Object {
              "history": Array [
                Object {
                  "date": 2017-07-07T00:00:00.000Z,
                  "value": 55.3,
                },
              ],
              "name": "Some name",
              "ownedAmount": 7,
              "value": 33.3,
            }
          }
          cancelFunc={[Function]}
          investFunc={[Function]}
        >
          <div />
        </mockConstructor>
      </InvestPanelRelay>
    </Relay(InvestPanelRelay)>
  </div>
</CtxProvider>
`;

exports[`InvestPanelRelay renders 2`] = `
Array [
  null,
  null,
  [Function],
]
`;

exports[`InvestPanelRelay renders 3`] = `
Array [
  RelayModernEnvironment {
    "__setNet": [Function],
    "_debugger": EnvironmentDebugger {
      "_envIsDirty": false,
      "_environment": [Circular],
      "_id": "RelayModernEnvironment1",
      "_isRecordingMutationEvents": false,
      "_recordedMutationEvents": Array [],
    },
    "_network": [Function],
    "_publishQueue": RelayPublishQueue {
      "_appliedOptimisticUpdates": Set {},
      "_backup": Object {},
      "_handlerProvider": [Function],
      "_pendingBackupRebase": false,
      "_pendingData": Set {},
      "_pendingOptimisticUpdates": Set {},
      "_pendingUpdaters": Set {},
      "_store": RelayMarkSweepStore {
        "_hasScheduledGC": false,
        "_index": 0,
        "_recordSource": Object {},
        "_roots": Map {},
        "_subscriptions": Set {
          Object {
            "callback": [Function],
            "snapshot": Object {
              "data": undefined,
              "dataID": "1",
              "node": Object {
                "argumentDefinitions": Array [],
                "kind": "Fragment",
                "metadata": null,
                "name": "InvestPanelRelay_bucket",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "id",
                    "storageKey": null,
                  },
                  Object {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "name",
                    "storageKey": null,
                  },
                  Object {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "value",
                    "storageKey": null,
                  },
                  Object {
                    "alias": null,
                    "args": null,
                    "kind": "ScalarField",
                    "name": "ownedAmount",
                    "storageKey": null,
                  },
                  Object {
                    "alias": null,
                    "args": Array [
                      Object {
                        "kind": "Literal",
                        "name": "count",
                        "type": "Int",
                        "value": 30,
                      },
                    ],
                    "concreteType": "GDataPoint",
                    "kind": "LinkedField",
                    "name": "history",
                    "plural": true,
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "date",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "value",
                        "storageKey": null,
                      },
                    ],
                    "storageKey": "history{\\"count\\":30}",
                  },
                ],
                "type": "GInvestmentBucket",
              },
              "seenRecords": Object {
                "1": undefined,
              },
              "variables": Object {},
            },
          },
          Object {
            "callback": [Function],
            "snapshot": Object {
              "data": undefined,
              "dataID": "2",
              "node": Object {
                "argumentDefinitions": Array [],
                "kind": "Fragment",
                "metadata": null,
                "name": "InvestPanelRelay_profile",
                "selections": Array [
                  Object {
                    "alias": null,
                    "args": null,
                    "concreteType": "GTradingAccount",
                    "kind": "LinkedField",
                    "name": "selectedAcc",
                    "plural": false,
                    "selections": Array [
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "availableCash",
                        "storageKey": null,
                      },
                      Object {
                        "alias": null,
                        "args": null,
                        "kind": "ScalarField",
                        "name": "id",
                        "storageKey": null,
                      },
                    ],
                    "storageKey": null,
                  },
                ],
                "type": "GProfile",
              },
              "seenRecords": Object {
                "2": undefined,
              },
              "variables": Object {},
            },
          },
        },
        "_updatedRecordIDs": Object {},
      },
    },
    "_store": RelayMarkSweepStore {
      "_hasScheduledGC": false,
      "_index": 0,
      "_recordSource": Object {},
      "_roots": Map {},
      "_subscriptions": Set {
        Object {
          "callback": [Function],
          "snapshot": Object {
            "data": undefined,
            "dataID": "1",
            "node": Object {
              "argumentDefinitions": Array [],
              "kind": "Fragment",
              "metadata": null,
              "name": "InvestPanelRelay_bucket",
              "selections": Array [
                Object {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "id",
                  "storageKey": null,
                },
                Object {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "name",
                  "storageKey": null,
                },
                Object {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "value",
                  "storageKey": null,
                },
                Object {
                  "alias": null,
                  "args": null,
                  "kind": "ScalarField",
                  "name": "ownedAmount",
                  "storageKey": null,
                },
                Object {
                  "alias": null,
                  "args": Array [
                    Object {
                      "kind": "Literal",
                      "name": "count",
                      "type": "Int",
                      "value": 30,
                    },
                  ],
                  "concreteType": "GDataPoint",
                  "kind": "LinkedField",
                  "name": "history",
                  "plural": true,
                  "selections": Array [
                    Object {
                      "alias": null,
                      "args": null,
                      "kind": "ScalarField",
                      "name": "date",
                      "storageKey": null,
                    },
                    Object {
                      "alias": null,
                      "args": null,
                      "kind": "ScalarField",
                      "name": "value",
                      "storageKey": null,
                    },
                  ],
                  "storageKey": "history{\\"count\\":30}",
                },
              ],
              "type": "GInvestmentBucket",
            },
            "seenRecords": Object {
              "1": undefined,
            },
            "variables": Object {},
          },
        },
        Object {
          "callback": [Function],
          "snapshot": Object {
            "data": undefined,
            "dataID": "2",
            "node": Object {
              "argumentDefinitions": Array [],
              "kind": "Fragment",
              "metadata": null,
              "name": "InvestPanelRelay_profile",
              "selections": Array [
                Object {
                  "alias": null,
                  "args": null,
                  "concreteType": "GTradingAccount",
                  "kind": "LinkedField",
                  "name": "selectedAcc",
                  "plural": false,
                  "selections": Array [
                    Object {
                      "alias": null,
                      "args": null,
                      "kind": "ScalarField",
                      "name": "availableCash",
                      "storageKey": null,
                    },
                    Object {
                      "alias": null,
                      "args": null,
                      "kind": "ScalarField",
                      "name": "id",
                      "storageKey": null,
                    },
                  ],
                  "storageKey": null,
                },
              ],
              "type": "GProfile",
            },
            "seenRecords": Object {
              "2": undefined,
            },
            "variables": Object {},
          },
        },
      },
      "_updatedRecordIDs": Object {},
    },
    "configName": undefined,
    "unstable_internal": Object {
      "areEqualSelectors": [Function],
      "createFragmentSpecResolver": [Function],
      "createOperationSelector": [Function],
      "getDataIDsFromObject": [Function],
      "getFragment": [Function],
      "getOperation": [Function],
      "getSelector": [Function],
      "getSelectorList": [Function],
      "getSelectorsFromObject": [Function],
      "getVariablesFromObject": [Function],
    },
  },
]
`;

exports[`InvestPanelRelay renders 4`] = `
Array [
  Object {
    "bucketId": "1",
    "quantity": 3,
    "tradingAccId": "3",
  },
  Object {
    "invest": Object {
      "bucket": Object {
        "id": "1",
        "ownedAmount": 10,
      },
      "tradingAccount": Object {
        "availableCash": 1134.6599999999999,
        "id": "3",
      },
    },
  },
]
`;

exports[`InvestPanelRelay renders 5`] = `
Array [
  Array [
    Object {
      "message": "Successfully bought the bucket!",
    },
  ],
  Array [
    Object {
      "message": "Successfully sold the bucket!",
    },
  ],
  Array [
    Object {
      "message": "Test",
    },
  ],
]
`;
